src/c++/perf_analyzer/test_command_line_parser.cc

```diff
@@ -0,0 +1,1040 @@
+
+    PAParamsPtr act;
+    REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+    REQUIRE(!parser.UsageCalled());
+
+    CHECK_PARAMS(act, exp);
+    optind = 1;
+  }
+
+  SUBCASE("Option : --streaming")
+  {
+    SUBCASE("streaming option - without model")
+    {
+      int argc = 2;
+      char* argv[argc] = {app_name, "--streaming"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(parser.UsageCalled());
+      CHECK_STRING(
+          "Usage Message", parser.GetUsageMessage(),
+          "streaming is only allowed with gRPC protocol");
+
+      exp->model_name = "";
+      exp->streaming = true;
+      // exp->max_threads = 16;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("with model")
+    {
+      int argc = 4;
+      char* argv[argc] = {app_name, "-m", model_name, "--streaming"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(parser.UsageCalled());
+
+      // NOTE: This is not an informative error message, how do I specify a gRPC
+      // protocol? Error ouput should list missing params.
+      //
+      CHECK_STRING(
+          "Usage Message", parser.GetUsageMessage(),
+          "streaming is only allowed with gRPC protocol");
+
+      exp->streaming = true;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("with model last")
+    {
+      int argc = 4;
+      char* argv[argc] = {app_name, "--streaming", "-m", model_name};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+
+      REQUIRE(parser.UsageCalled());
+      CHECK_STRING(
+          "Usage Message", parser.GetUsageMessage(),
+          "streaming is only allowed with gRPC protocol");
+
+      exp->streaming = true;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+  }
+
+  SUBCASE("Option : --max-threads")
+  {
+    SUBCASE("set to 1")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--max-threads", "1"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(!parser.UsageCalled());
+
+      exp->max_threads = 1;
+      exp->max_threads_specified = true;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("set to max")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--max-threads", "65535"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(!parser.UsageCalled());
+
+      exp->max_threads = 65535;
+      exp->max_threads_specified = true;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("missing value")
+    {
+      int argc = 4;
+      char* argv[argc] = {app_name, "-m", model_name, "--max-threads"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(parser.UsageCalled());
+
+      // NOTE: Empty message is not helpful
+      //
+      CHECK_STRING("Usage Message", parser.GetUsageMessage(), "");
+      // BUG: Dumping string "option '--max-threads' requires an argument"
```

Comment:
That string is coming from the getopt library. You can set `opterr = 0` at the beginning of this `TEST_CASE()` to disable the getopt library from printing error messages.

https://www.gnu.org/software/libc/manual/html_node/Using-Getopt.html

Reply:
This is one of the things that I plan to refactor. I would like to remove the use of getopt, as it is antiquated and not really suitable for modern C++ applications. The note is to help other understand the behavior and will get addressed.

Reply:
> This is one of the things that I plan to refactor.

Are you intending to fix the bug where the getopt library prints error messages during unit testing in this PR?

Reply:
I wasn't planning on it. But if they get to be too many, then I might. I like to have them print out in the unit test output to remind me that I need to fix it. If I disable it, I am apt to forget about it.

Reply:
I think the error message should be disabled to keep the unit test output clean between now and when the getopt library is removed. A ticket can be created for removing the getopt library.

Reply:
I went ahead and disabled it for now. The problem I have with this is it masks a behavior. Perhaps it is incorrect, but it is now hidden from the Unit Tests. Meaning, if I refactor, and something else it expecting this output, the refactor could change the behavior. This is why I was reluctant to disable it.

Reply:
You could set `oldopterr = opterr; opterr = 0` at the beginning of the `TEST_CASE()` polluting the unit test output and then set `opterr = oldopterr` at the end of the `TEST_CASE()`, therefore preventing the situation where something else is expecting the output and can't get it.

Reply:
No, that is too obscure, but thank you for the suggestion.